testutil/ibctesting/chain.go

```diff
@@ -1,800 +1,126 @@
+	cmttypes "github.com/cometbft/cometbft/types"
 
-	DefaultDelayPeriod uint64 = 0
+	ibctesting "github.com/cosmos/ibc-go/v8/testing"
 )
 
-var (
-	DefaultOpenInitVersion *connectiontypes.Version
-	DefaultTrustLevel      = ibctmtypes.DefaultTrustLevel
-	UpgradePath            = []string{"upgrade", "upgradedIBCState"}
-	ConnectionVersion      = connectiontypes.ExportedVersionsToProto(connectiontypes.GetCompatibleVersions())[0]
-)
-
-// TestChain is a testing struct that wraps a simapp with the last TM Header, the current ABCI
-// header and the validators of the TestChain. It also contains a field called ChainID. This
-// is the clientID that *other* chains use to refer to this TestChain. The SenderAccount
-// is used for delivering transactions through the application state.
-// NOTE: the actual application uses an empty chain-id for ease of testing.
-type TestChain struct {
-	T *testing.T
-
-	App           *simapp.DesmosApp
-	ChainID       string
-	LastHeader    *ibctmtypes.Header // header for last block height committed
-	CurrentHeader tmproto.Header     // header for current block height
-	QueryServer   types.QueryServer
-	TxConfig      client.TxConfig
-	Codec         codec.BinaryCodec
-
-	Vals    *tmtypes.ValidatorSet
-	Signers []tmtypes.PrivValidator
-
-	PrivKey cryptotypes.PrivKey
-	Account authtypes.AccountI
-
-	// IBC specific helpers
-	ClientIDs   []string          // ClientID's used on this chain
-	Connections []*TestConnection // track connectionID's created for this chain
-}
-
-// NewTestChain initializes a new TestChain instance with a single validator set using a
-// generated private key. It also creates a sender account to be used for delivering transactions.
+// NewTestChainWithValSet initializes a new TestChain instance with the given validator set
+// and signer array. It also initializes 10 Sender accounts with a balance of 10000000000000000000 coins of
+// bond denom to use for tests.
 //
 // The first block height is committed to state in order to allow for client creations on
 // counterparty chains. The TestChain will return with a block height starting at 2.
 //
 // Time management is handled by the Coordinator in order to ensure synchrony between chains.
 // Each update of any chain increments the block header time for all chains by 5 seconds.
-func NewTestChain(t *testing.T, chainID string) *TestChain {
-	// generate validator private/public key
-	privVal := mock.NewPV()
-	pubKey, err := privVal.GetPubKey()
-	require.NoError(t, err)
+//
+// NOTE: to use a custom sender privkey and account for testing purposes, replace and modify this
+// constructor function.
+//
+// CONTRACT: Validator array must be provided in the order expected by Tendermint.
+// i.e. sorted first by power and then lexicographically by address.
+func NewTestChainWithValSet(tb testing.TB, coord *ibctesting.Coordinator, chainID string, valSet *cmttypes.ValidatorSet, signers map[string]cmttypes.PrivValidator) *ibctesting.TestChain {
```

Comment:
FinalizeBlock phase is introduced in comebft 0.38.0, it breaks the ibctesting tool, so I decided to wrap the existing tool from ibc-go, instead of maintaining it by ourselves.